### PR TITLE
Add documentation for Routing surface area

### PR DESCRIPTION
**Pull request by captainsafia**
Part of #26397.

### CODE REVIEW

File: src/Http/Routing/src/RouteBuilder.cs [hunk 1]
```diff
@@ -10,13 +10,26 @@
...
 {
+    /// <summary>
+    /// Provides support for specifying routes in an application.
+    /// </summary>
     public class RouteBuilder : IRouteBuilder
     {
+        /// <summary>
+        /// Constructs a new <see cref="RouteBuilder"/> instance given an <paramref name="applicationBuilder"/>.
+        /// </summary>
+        /// <param name="applicationBuilder">A <see cref="IApplicationBuilder"/> instance.</param>
```
Comment (JamesNK) [inline src/Http/Routing/src/RouteBuilder.cs:21]:
```suggestion
        /// <param name="applicationBuilder">An <see cref="IApplicationBuilder"/> instance.</param>
```

File: src/Http/Routing/src/RouteBuilder.cs [hunk 2]
```diff
@@ -10,13 +10,26 @@
...
         public RouteBuilder(IApplicationBuilder applicationBuilder)
             : this(applicationBuilder, defaultHandler: null)
         {
         }
 
+        /// <summary>
+        /// Constructs a new <see cref="RouteBuilder"/> instance given an <paramref name="applicationBuilder"/>
+        /// and <paramref name="defaultHandler"/>.
+        /// </summary>
+        /// <param name="applicationBuilder">A <see cref="IApplicationBuilder"/> instance.</param>
```
Comment (JamesNK) [inline src/Http/Routing/src/RouteBuilder.cs:31]:
```suggestion
        /// <param name="applicationBuilder">An <see cref="IApplicationBuilder"/> instance.</param>
```

File: src/Http/Routing/src/RouteCollection.cs [hunk 3]
```diff
@@ -24,16 +27,24 @@ public class RouteCollection : IRouteCollection
 
         private RouteOptions? _options;
 
+        /// <summary>
+        /// Gets the route at a given index.
+        /// </summary>
+        /// <value></value>
```
Comment (JamesNK) [inline src/Http/Routing/src/RouteCollection.cs:33]:
```suggestion
        /// <value>The route at the given index.</value>
```

File: src/Http/Routing/src/RouterMiddleware.cs [hunk 4]
```diff
@@ -9,12 +9,21 @@
...
+    /// Middleware responsible for routing.
+    /// </summary>
     public class RouterMiddleware
     {
         private readonly ILogger _logger;
         private readonly RequestDelegate _next;
         private readonly IRouter _router;
 
+        /// <summary>
+        /// Constructor for <see cref="RouterMiddleware"/> with a given <paramref name="router"/>.
```
Comment (JamesNK) [inline src/Http/Routing/src/RouterMiddleware.cs:22]:
```suggestion
        /// Constructs a new <see cref="RouterMiddleware"/> instance with a given <paramref name="router"/>.
```

File: src/Http/Routing/src/Template/InlineConstraint.cs [hunk 5]
```diff
@@ -25,6 +25,10 @@ public InlineConstraint(string constraint)
             Constraint = constraint;
         }
 
+        /// <summary>
+        /// Creates a new instance of <see cref="InlineConstraint"/> given a <see cref="RoutePatternParameterPolicyReference"/>.
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/InlineConstraint.cs:29]:
```suggestion
        /// Creates a new <see cref="InlineConstraint"/> instance given a <see cref="RoutePatternParameterPolicyReference"/>.
```

File: src/Http/Routing/src/Template/RouteTemplate.cs [hunk 6]
```diff
@@ -11,11 +11,18 @@
...
+    /// <summary>
+    /// Represents the template for a route.
+    /// </summary>
     [DebuggerDisplay("{DebuggerToString()}")]
     public class RouteTemplate
     {
         private const string SeparatorString = "/";
 
+        /// <summary>
+        /// Creates a new <see cref="RouteTemplate"/> given <paramref name="other"/>.
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/RouteTemplate.cs:23]:
```suggestion
        /// Constructs a new <see cref="RouteTemplate"/> instance given <paramref name="other"/>.
```

File: src/Http/Routing/src/Template/RouteTemplate.cs [hunk 7]
```diff
@@ -42,6 +49,12 @@ public RouteTemplate(RoutePattern other)
             }
         }
 
+        /// <summary>
+        /// Creates a new route template instance given the <paramref name="template"/> string
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/RouteTemplate.cs:53]:
```suggestion
        /// Constructs a a new <see cref="RouteTemplate" /> instance given the <paramref name="template"/> string
```

File: src/Http/Routing/src/Template/RouteTemplate.cs [hunk 8]
```diff
@@ -68,12 +81,27 @@ public RouteTemplate(string template, List<TemplateSegment> segments)
             }
         }
 
+        /// <summary>
+        /// Gets the strign representation of the route template.
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/RouteTemplate.cs:85]:
```suggestion
        /// Gets the string representation of the route template.
```

File: src/Http/Routing/src/Template/RouteTemplate.cs [hunk 9]
```diff
@@ -68,12 +81,27 @@ public RouteTemplate(string template, List<TemplateSegment> segments)
...
+        /// </summary>
         public string? TemplateText { get; }
 
+        /// <summary>
+        /// Gets the list of <see cref="TemplatePart"/> that represent that parameters defined in the route template.
+        /// </summary>
+        /// <value></value>
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/RouteTemplate.cs:92]: Remove?

File: src/Http/Routing/src/Template/TemplatePart.cs [hunk 10]
```diff
@@ -10,13 +10,23 @@
...
 namespace Microsoft.AspNetCore.Routing.Template
 {
+    /// <summary>
+    /// Represents a part of a route template segment.
+    /// </summary>
     [DebuggerDisplay("{DebuggerToString()}")]
     public class TemplatePart
     {
+        /// <summary>
+        /// Constructor for a new <see cref="TemplatePart"/> instance.
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/TemplatePart.cs:20]:
```suggestion
        /// Constructs a new <see cref="TemplatePart"/> instance.
```

File: src/Http/Routing/src/Template/TemplatePart.cs [hunk 11]
```diff
@@ -10,13 +10,23 @@
...
+        /// </summary>
         public TemplatePart()
         {
         }
 
+        /// <summary>
+        /// Creates a <see cref="TemplatePart"/> instance given a <paramref name="other"/>.
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/TemplatePart.cs:27]:
```suggestion
        /// Constructs a new <see cref="TemplatePart"/> instance given a <paramref name="other"/>.
```

File: src/Http/Routing/src/Template/TemplateSegment.cs [hunk 12]
```diff
@@ -9,14 +9,24 @@
...
 namespace Microsoft.AspNetCore.Routing.Template
 {
+    /// <summary>
+    /// Represents a segment of a route template.
+    /// </summary>
     [DebuggerDisplay("{DebuggerToString()}")]
     public class TemplateSegment
     {
+        /// <summary>
+        /// Constructor for a new <see cref="TemplateSegment"/>.
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/TemplateSegment.cs:19]:
```suggestion
        /// Constructs a new <see cref="TemplateSegment"/> instance.
```

File: src/Http/Routing/src/Template/TemplateSegment.cs [hunk 13]
```diff
@@ -9,14 +9,24 @@
...
+        /// </summary>
         public TemplateSegment()
         {
             Parts = new List<TemplatePart>();
         }
 
+        /// <summary>
+        /// Constructor for a new <see cref="TemplateSegment"/> given another <see cref="RoutePatternPathSegment"/>.
```
Comment (JamesNK) [inline src/Http/Routing/src/Template/TemplateSegment.cs:27]:
```suggestion
        /// Constructs a new <see cref="TemplateSegment"/> instance given another <see cref="RoutePatternPathSegment"/>.
```

**Review by JamesNK — APPROVED** [review]
Just a bunch of minor things.

File: src/Http/Routing/src/Matching/INodeBuilderPolicy.cs [hunk 14]
```diff
@@ -6,12 +6,31 @@
 
 namespace Microsoft.AspNetCore.Routing.Matching
 {
+    /// <summary>
+    /// Implements an interface for a matcher policy with support for generating graph representations of the endpoints.
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/Routing/src/Matching/INodeBuilderPolicy.cs:10]: That would be great! I tried my best to grok what was going on here but quite a few of the concepts are new to me.

File: src/Http/Routing/src/Template/RouteTemplate.cs (hunk 9 above)
Comment (captainsafia) [reply src/Http/Routing/src/Template/RouteTemplate.cs:92]: Good catch!

**Pull request merged**